lsp-clients.el

```diff
@@ -305,37 +305,29 @@ finding the executable with `exec-path'."
 
 ;; Dart
 (defgroup lsp-dart nil
-  "LSP support for Dart, using dart_language_server."
+  "LSP support for Dart, using dart analysis server."
   :group 'lsp-mode
-  :link '(url-link "https://github.com/natebosch/dart_language_server"))
+  :link '(url-link "https://github.com/dart-lang/sdk/tree/master/pkg/analysis_server"))
 
-(defcustom lsp-clients-dart-server-command
-  (expand-file-name (if (equal system-type 'windows-nt)
-                        "~/Pub/Cache/bin/dart_language_server"
-                      "~/.pub-cache/bin/dart_language_server"))
-  "The dart_language_server executable to use."
+(defcustom lsp-dart-sdk-dir "~/flutter/bin/cache/dart-sdk/"
+  "Install directory for dart-sdk. A slash is expected at the end."
   :group 'lsp-dart
-  :type 'file)
+  :risky t
+  :type 'directory)
 
-(defun lsp-dart--lsp-command ()
+(defun lsp-dart--analysis-server-command ()
   "Generate LSP startup command."
-  (let ((dls lsp-clients-dart-server-command)
-        (pub (executable-find "pub")))
-    (if pub
-        (if (executable-find dls)
-            dls
-          (message "Installing dart_language_server...")
-          (shell-command (concat pub " global activate dart_language_server"))
-          (message "Installed dart_language_server")
-          dls)
-      (error "Please ensure /path/to/dart-sdk/bin is on system path"))))
+  `("dart"
+    ,(expand-file-name (concat lsp-dart-sdk-dir "bin/snapshots/analysis_server.dart.snapshot"))
```

Comment:
use f-join instead of concat and fix the comment above(no slash will be required after this change.